### ISSUE TITLE
Saved NER model is different from the NER model before saved

### ISSUE THREAD

**Issue by chunlei2** [explosion/spaCy#4649]
## How to reproduce the behaviour

Adjust the example code in [Spacy documentation](https://spacy.io/usage/training#ner)  which trains an additional entity type `animal`

Replace `LABEL =  "ANIMAL"` with `LABEL = "DATE"`

Replace `TRAIN_DATA` with `TRAIN_DATA = [
    (
        "ADHD (2008)",
        {"entities": [(6, 10, LABEL)]}
    )
]` 

Replace `test_text = "Do you like horses?"` with `test_text = "ADHD (2008)"`

Call main function `main(model = 'en_core_web_sm', output_dir='test')`

The printed output will be `Entities in 'ADHD (2008)'
Saved model to test
Loading from test
VEGETABLE 2008)`
which means the trained model doesn't detect any entity, but the saved model detect VEGETABLE entity. And VEGETABLE shouldn't appear because training data doesn't contain the entity type.

Thanks in advance!
## Your Environment

* Operating System: Windows Server 2012 R2
* Python Version Used: 3.6.4
* spaCy Version Used: 2.1.8
* Environment Information:

**Comment by adrianeboyd** [comment]
Hi, thanks for the report! Can you provide the full script that you are using so it's easier for us to see if we can replicate the problem?

**Comment by chunlei2** [comment]
Hi, the code is attached. I also change dropout from 0.35 to 0, make sure every time outputting same result.
``` python
from __future__ import unicode_literals, print_function

import plac
import random
from pathlib import Path
import spacy
from spacy.util import minibatch, compounding


# new entity label
LABEL = "DATE"

# training data
# Note: If you're using an existing model, make sure to mix in examples of
# other entity types that spaCy correctly recognized before. Otherwise, your
# model might learn the new type, but "forget" what it previously knew.
# https://explosion.ai/blog/pseudo-rehearsal-catastrophic-forgetting
TRAIN_DATA = [ ( "ADHD (2008)", {"entities": [(6, 10, LABEL)]} ) ]


def main(model=None, new_model_name="animal", output_dir=None, n_iter=30):
    """Set up the pipeline and entity recognizer, and train the new entity."""
    random.seed(0)
    if model is not None:
        nlp = spacy.load(model)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("en")  # create blank Language class
        print("Created blank 'en' model")
    # Add entity recognizer to model if it's not in the pipeline
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if "ner" not in nlp.pipe_names:
        ner = nlp.create_pipe("ner")
        nlp.add_pipe(ner)
    # otherwise, get it, so we can add labels to it
    else:
        ner = nlp.get_pipe("ner")

    ner.add_label(LABEL)  # add new entity label to entity recognizer
    # Adding extraneous labels shouldn't mess anything up
    ner.add_label("VEGETABLE")
    if model is None:
        optimizer = nlp.begin_training()
    else:
        optimizer = nlp.resume_training()
    move_names = list(ner.move_names)
    # get names of other pipes to disable them during training
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != "ner"]
    with nlp.disable_pipes(*other_pipes):  # only train NER
        sizes = compounding(1.0, 4.0, 1.001)
        # batch up the examples using spaCy's minibatch
        for itn in range(n_iter):
            random.shuffle(TRAIN_DATA)
            batches = minibatch(TRAIN_DATA, size=sizes)
            losses = {}
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(texts, annotations, sgd=optimizer, drop=0, losses=losses)
            print("Losses", losses)

    # test the trained model
    test_text = "ADHD (2008)"
    doc = nlp(test_text)
    print("Entities in '%s'" % test_text)
    for ent in doc.ents:
        print(ent.label_, ent.text)

    # save model to output directory
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.meta["name"] = new_model_name  # rename model
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        # Check the classes have loaded back consistently
        assert nlp2.get_pipe("ner").move_names == move_names
        doc2 = nlp2(test_text)
        for ent in doc2.ents:
            print(ent.label_, ent.text)
main(model = 'en_core_web_sm', output_dir='test')
```

**Comment by ines** [comment]
I just ran your script with the latest version and can't reproduce the problem – I'm seeing `DATE 2008` as the only entity. It's possible that what you're seeing here is a problem with the model in v2.1.x that was since fixed in v2.2.

Also note that the way you're running your script here isn't that helpful to really determine the model results. You're updating an existing model that already predicts `DATE` with one single example. So whatever you get out at the end really isn't that representative.

**Comment by chunlei2** [comment]
Hi, thanks for the note. I don't mean to create a representative model, just want to figure out why the saved model is different from the model before saved. I upgrade the spacy to 2.2.2 and still have the issue. I reproduce the error using Google Colab.[Link](https://colab.research.google.com/drive/1vOsDU_7NZ1FCLy3E22eU-Y9qIF4IB3uw)

**Comment by adrianeboyd** [comment]
I could reproduce the problem, so I'm going to reopen this issue. It's possible it's just spurious behavior for labels without enough data, but the save/load behavior is indeed unexpected.

**Comment by honnibal** [comment]
If this occurs in v2.2, then yes I would say it's a bug. We're supposed to have a filter in the model that prevents prediction of classes that haven't been seen in any training examples.

**Comment by adrianeboyd** [comment]
I checked again and I only see unusual results with v2.1.9, not v2.2.3.

**Comment by chunlei2** [comment]
Hi, I run that google colab link and the problem still exists for v2.2.3.

**Comment by svlandeg** [comment]
I'm afraid I also can't reproduce this in v2.2.3, so it's kind of unclear what is going on.

@chunlei2: could you make sure in your script, that the output directory is entirely empty before you save to it? Also, can you reproduce this problem in a script run on your local PC - i.e. outside of google colab? And can you run the Google Colab with 2.2.3 instead of 2.2.2 ?
